File: web/src/components/products/ProductService.js
import http from '../http'

const api = {
    getProducts: getProducts,
    getProduct: getProduct,
    save: save,
    deleteProduct: deleteProduct
};

const productsUrl = 'http://localhost:9999/kona/products';

function save(product) {
    if (product.id > 0) {
        return http.PUT(`${productsUrl}/${product.id}`, product);
    } else {
        return http.POST(productsUrl, product);
    }
}

function getProducts(name) {

    const queryParams = {};

    if (name) {
        queryParams.name = name;
    }

    return http.GET(productsUrl, queryParams);
}

function deleteProduct(productId) {
    return http.DELETE(`${productsUrl}/${productId}`)
}

function getProduct(productId) {
    return http.GET(`${productsUrl}/${productId}`);
}

export default api;

File: application-services/src/main/java/kona/application/services/ProductService.java
package kona.application.services;

import kona.domain.product.Product;
import kona.domain.product.ProductRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class ProductService {

    private final ProductRepository productRepository;

    public ProductService(final ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    @Transactional(readOnly = true)
    public List<Product> findBy(String name) {
        if (name == null) {
            return productRepository.getAll();
        } else {
            return productRepository.findBy(name);
        }
    }

    @Transactional
    public void save(Product product) {
        productRepository.add(product);
    }

    @Transactional(readOnly = true)
    public Product findBy(long id) {
        return productRepository
                .findBy(id)
                .orElseThrow(() -> new NotFoundException(Product.class, "id="+id));
    }

    @Transactional
    public void update(Product product) {
        productRepository.update(product);
    }

    @Transactional
    public void delete(long productId) {
        int deletedRows = productRepository.delete(productId);
        if (deletedRows == 0) {
            throw new NotFoundException(Product.class, "id="+productId);
        }
    }
}
